### PR TITLE
Example for size function

### LINKED ISSUE #387: Galaxy Size example
## Description
To create an example to reproduce figure 4 in Shen et al. 2003. 
This example will have the same format as the luminosity example in PR #385.

## References
https://arxiv.org/pdf/astro-ph/0301527.pdf

### PR DESCRIPTION
**Pull request by Lucia-Fonseca**
## Description
This PR creates an example to reproduce figure 4 in Shen et al. 2003. It closes #387 
## Checklist
- [x] Follow the [Contributor Guidelines](https://github.com/skypyproject/skypy/blob/master/CONTRIBUTING.md)
- [ ] Write unit tests
- [x] Write documentation strings
- [x] Assign someone from your working team to review this pull request
- [x] Assign someone from the infrastructure team to review this pull request

## Help
I need help with the SDSS dataset.

### CODE REVIEW

**Comment by Lucia-Fonseca** [comment]
Help: I updated the branch after #385 was merged but I cannot see the example in the docs created by this PR.
What am I missing?

**Comment by Lucia-Fonseca** [comment]
Ready for review

File: examples/galaxies/plot_size.py [hunk 1]
```diff
@@ -0,0 +1,144 @@
...
+# Load SDSS data release 7 and split the sample
+R50_r_phys, c, M_r = np.genfromtxt('SDSS_DR7.csv', delimiter=',')
+M_r_late, R50_r_phys_late = M_r[c < 2.86], R50_r_phys[c < 2.86]
+M_r_early, R50_r_phys_early = M_r[c > 2.86], R50_r_phys[c > 2.86]
+
+# Load data from figure 4 in Shen et al 2003 and correct for shift in bins
+sdss_early = np.loadtxt('Shen+03_early.txt')
+sdss_late = np.loadtxt('Shen+03_late.txt')
+sdss_early[:, 0] -= 0.25
+sdss_late[:, 0] -= 0.25
```
Comment (ntessore) [inline examples/galaxies/plot_size.py:96]: the correct numbers should be in the data files

File: examples/galaxies/plot_size.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+# Size-magnitude relation
+# -------------------------
+#
+# According to Shen et al. 2003 [1]_, the observed size of galaxies
+# and the absolute magnitudes follow simple analytic formulae. For early-type
+# galaxies, the relation (14) reads
+#
+# .. math::
+#
+#    log (\bar{R}/kpc) = -0.4aM + b,
```
Comment (aamara) [inline examples/galaxies/plot_size.py:21]: Do we want to be specific about the base of the log? I'm guessing log_10 rather than ln. Only a minor thought.

Comment (ntessore) [inline examples/galaxies/plot_size.py]:
```suggestion
#    \log (\bar{R}/{\rm kpc}) = -0.4aM + b,
```

File: examples/galaxies/plot_size.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+#
+# with :math:`a` and :math:`b` fitting constants. Likewise, for late-type galaxies
+# the formula (15) reads
+#
+# .. math::
+#
+#    log (\bar{R}/kpc) = -0.4\alpha M +(\beta -\alpha) log \left[1 + 10^{-0.4(M-M_0)}\right] +\gamma
```
Comment (ntessore) [inline examples/galaxies/plot_size.py]:
```suggestion
#    \log (\bar{R}/{\rm kpc}) = -0.4\alpha M +(\beta -\alpha) \log \left[1 + 10^{-0.4(M-M_0)}\right] +\gamma
```

File: examples/galaxies/plot_size.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+# .. math::
+#
+#    p(R) = \frac{1}{\sigma R \sqrt{2 \pi}}e^{- \frac{(ln R - \mu)^2}{2 \sigma^2}}
+#
+# with mean :math:`\mu = 0` and standard deviation :math:`\sigma`, and an amplitude
+# :math:`\bar{R}` given by the Shen et al. 2003 model:
+# :func:`skypy.galaxy.size.early_type_lognormal()` and
+# :func:`skypy.galaxy.size.late_type_lognormal()`
+# (c.f. equations above).
+#
```
Comment (ntessore) [inline examples/galaxies/plot_size.py:50]: Not necessary

File: examples/galaxies/plot_size.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+# Parameters for the late-type and early-type galaxies
+alpha, beta, gamma = 0.21, 0.53, -1.31
+a, b = 0.6, -4.63
+M0 = -20.52
+sigma1, sigma2 = 0.48, 0.25
+
+# SkyPy late sample
+M_bins_late = np.arange(-16, -24.1, -0.5)
+M_late = np.random.uniform(M_bins_late[0], M_bins_late[-1], size=138000)
+R_late = late_type_lognormal(M_late, alpha, beta, gamma, M0, sigma1, sigma2).value
```
Comment (ntessore) [inline examples/galaxies/plot_size.py:67]: for simplicity, I would use only the numbers here

```suggestion
M_late = np.random.uniform(-16, -24, size=10000)
R_late = late_type_lognormal(M_late, alpha, beta, gamma, M0, sigma1, sigma2).value
```

File: examples/galaxies/plot_size.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+
+# SkyPy early sample
+M_bins_early = np.arange(-18, -24.1, -0.5)
+M_early = np.random.uniform(M_bins_early[0], M_bins_early[-1], size=138000)
+R_early = early_type_lognormal(M_early, a, b, M0, sigma1, sigma2).value
```
Comment (ntessore) [inline examples/galaxies/plot_size.py:72]:
```suggestion
M_early = np.random.uniform(-18, -24, size=10000)
R_early = early_type_lognormal(M_early, a, b, M0, sigma1, sigma2).value
```

File: examples/galaxies/plot_size.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+# Load data from figure 4 in Shen et al 2003
+sdss_early = np.loadtxt('Shen+03_early.txt')
+sdss_late = np.loadtxt('Shen+03_late.txt')
+
+# Median sizes for SkyPy late- and early-type galaxies
+R_bar_early = [np.median(R_early[(M_early <= Ma) & (M_early > Mb)])
+               for Ma, Mb in zip(M_bins_early, M_bins_early[1:])]
+R_bar_late = [np.median(R_late[(M_late <= Ma) & (M_late > Mb)])
+              for Ma, Mb in zip(M_bins_late, M_bins_late[1:])]
```
Comment (ntessore) [inline examples/galaxies/plot_size.py:105]: then you can define the bins a little closer to where they are used

```suggestion

# Bins for median radii
M_bins_late = np.arange(-16, -24.1, -0.5)
M_bins_early = np.arange(-18, -24.1, -0.5)

# Median sizes for SkyPy late- and early-type galaxies
R_bar_early = [np.median(R_early[(M_early <= Ma) & (M_early > Mb)])
               for Ma, Mb in zip(M_bins_early, M_bins_early[1:])]
R_bar_late = [np.median(R_late[(M_late <= Ma) & (M_late > Mb)])
              for Ma, Mb in zip(M_bins_late, M_bins_late[1:])]
```

File: examples/galaxies/plot_size.py (hunk 4 above)
Comment (Lucia-Fonseca) [reply examples/galaxies/plot_size.py:51]: What is not necessary, the equation or the explanation?

Comment (ntessore) [reply examples/galaxies/plot_size.py:51]: neither is I think, that it is a lognormal distribution is literally the model

File: examples/galaxies/plot_size.py [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+# and `late-type
+# <https://github.com/skypyproject/skypy/raw/master/examples/galaxies/Shen+03_late.txt>`_.
+# The columns of these files represent magnitudes, radii, lower error and
+# upper error.
+
+# Load data from figure 4 in Shen et al 2003
+sdss_early = np.loadtxt('Shen+03_early.txt')
+sdss_late = np.loadtxt('Shen+03_late.txt')
+error_late = sdss_late[:, 3] - sdss_late[:, 2]
+error_early = sdss_early[:, 3] - sdss_early[:, 2]
```
Comment (ntessore) [inline examples/galaxies/plot_size.py:88]: the errors are symmetric in logspace, i.e. `central_value = sqrt(upper*lower)`, not in linear space

I would store the minus and plus errors, i.e. `[:, 2] - [:, 1]` and `[:, 3] - [:, 1]` in the file so there needs to be no further processing in the example

Comment (Lucia-Fonseca) [reply examples/galaxies/plot_size.py]: I am not sure if I am addressing this correctly

**Review by itrharrison — COMMENTED** [review]
See comments above with just a few minor wording changes.

**Comment by Lucia-Fonseca** [comment]
This is ready now

**Review by ntessore — DISMISSED** [review]

File: examples/galaxies/plot_size.py [hunk 9]
```diff
@@ -0,0 +1,125 @@
+"""
+Galaxy size
+===========
```
Comment (rrjbca) [inline examples/galaxies/plot_size.py:3]:
```suggestion
Galaxy Size Distributions
=========================
```

File: examples/galaxies/plot_size.py [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+
+This example demonstrate how to obtain sizes for early and late type galaxies
```
Comment (rrjbca) [inline examples/galaxies/plot_size.py:5]:
```suggestion
This example demonstrate how to sample sizes for early and late type galaxies
```

File: examples/galaxies/plot_size.py [hunk 11]
```diff
@@ -0,0 +1,125 @@
...
+in SkyPy.
+
+"""
+
+
+# %%
+# Size-magnitude relation
```
Comment (rrjbca) [inline examples/galaxies/plot_size.py:12]:
```suggestion
# Size-Magnitude Relation
```

File: examples/galaxies/plot_size.py [hunk 12]
```diff
@@ -0,0 +1,125 @@
...
+# -------------------------
+#
+# According to Shen et al. 2003 [1]_, the observed size of galaxies
+# and the absolute magnitudes follow simple analytic formulae. For early-type
+# galaxies, the relation (14) reads
```
Comment (rrjbca) [inline examples/galaxies/plot_size.py:17]:
```suggestion
# In Shen et al. 2003 [1]_ the observed sizes :math:`R` of galaxies
# were shown to follow simple analytic relations as a function of their absolute magnitudes :math:`M`. For early-type
# galaxies, their mean radius follows Equation 14:
```

File: examples/galaxies/plot_size.py [hunk 13]
```diff
@@ -0,0 +1,125 @@
...
+#
+# .. math::
+#
+#    \log_{10} (\bar{R}/{\rm kpc}) = -0.4aM + b,
+#
+# with :math:`a` and :math:`b` fitting constants. Likewise, for late-type galaxies
+# the equation (15) reads
```
Comment (rrjbca) [inline examples/galaxies/plot_size.py:24]:
```suggestion
# with :math:`a` and :math:`b` fitting constants. Likewise, late-type galaxies
# follow Equation 15:
```

File: examples/galaxies/plot_size.py [hunk 14]
```diff
@@ -0,0 +1,125 @@
...
+#
+# .. math::
+#
+#    \log_{10}(\bar{R}/{\rm kpc})=-0.4\alpha M+
+#                                  (\beta -\alpha)\log \left[1+10^{-0.4(M-M_0)}\right]+\gamma
+#
+# with a dispersion given by equation 16
```
Comment (rrjbca) [inline examples/galaxies/plot_size.py:31]:
```suggestion
# The dispersion on these relations is given by Equation 16:
```

File: examples/galaxies/plot_size.py [hunk 15]
```diff
@@ -0,0 +1,125 @@
...
+#
+# where :math:`\alpha`, :math:`\beta`, :math:`\gamma`, :math:`\sigma_1`, :math:`\sigma_2` and
+# :math:`M_0` are fitting parameters.
+#
+# In SkyPy, we draw physical sizes of both galaxy types from a lognormal distribution
+# with standard deviation, :math:`\sigma`, and amplitude,
+# :math:`\bar{R}`, given by the Shen et al. 2003 model:
+# :func:`skypy.galaxy.size.early_type_lognormal()` and
+# :func:`skypy.galaxy.size.late_type_lognormal()`
+# (c.f. equations above).
```
Comment (rrjbca) [inline examples/galaxies/plot_size.py:45]:
```suggestion
# In SkyPy, we can sample physical sizes for each galaxy type from lognormal distributions
# with standard deviation :math:`\sigma` and amplitude
# :math:`\bar{R}` using the functions
# :func:`skypy.galaxy.size.early_type_lognormal()` and
# :func:`skypy.galaxy.size.late_type_lognormal()`.
```

File: examples/galaxies/plot_size.py [hunk 16]
```diff
@@ -0,0 +1,125 @@
...
+#
+# In this example, the values of the fitting parameters
+# are taken from the model in Shen et al. 2003 [1].
```
Comment (rrjbca) [inline examples/galaxies/plot_size.py:48]:
```suggestion
# In this example, we simulate the sizes of galaxies with random magnitudes using the values for the parameters
# given in Shen et al. 2003 Table 1 [1]_ :
```

File: examples/galaxies/plot_size.py [hunk 17]
```diff
@@ -0,0 +1,125 @@
...
+# %%
+# Validation with SDSS data
+# -------------------------
+# In this example, we reproduce figure 4 in reference [1]_.
+# The data files are included here: `early-type
+# <https://github.com/skypyproject/skypy/raw/master/examples/galaxies/Shen+03_early.txt>`_
+# and `late-type
+# <https://github.com/skypyproject/skypy/raw/master/examples/galaxies/Shen+03_late.txt>`_.
+# The columns of these files represent magnitudes, radii, minus error and
+# plus error.
```
Comment (rrjbca) [inline examples/galaxies/plot_size.py:77]:
```suggestion
# Validation against SDSS Data
# ----------------------------
# Here we reproduce Figure 4 from [1]_, comparing our simulated galaxy sizes against observational data from SDSS.
# You can download the data files for `early-type
# <https://github.com/skypyproject/skypy/raw/master/examples/galaxies/Shen+03_early.txt>`_
# and `late-type
# <https://github.com/skypyproject/skypy/raw/master/examples/galaxies/Shen+03_late.txt>`_.
# distributions which have the following columns: magnitudes, median radius, minus error, and
# plus error.
```

**Comment by Lucia-Fonseca** [comment]
This is now ready.

File: examples/galaxies/plot_size.py [hunk 18]
```diff
@@ -0,0 +1,126 @@
...
+# %%
+# Validation against SDSS Data
+# ----------------------------
+# Here we reproduce Figure 4 from [1]_, comparing our simulated galaxy sizes against
+# observational data from SDSS.
+# You can download the data files for `early-type
+# <https://github.com/skypyproject/skypy/raw/master/examples/galaxies/Shen+03_early.txt>`_
+# and `late-type
+# <https://github.com/skypyproject/skypy/raw/master/examples/galaxies/Shen+03_late.txt>`_.
+# distributions which have the following columns: magnitudes, median radius, minus error, and
```
Comment (rrjbca) [inline examples/galaxies/plot_size.py:77]:
```suggestion
# <https://github.com/skypyproject/skypy/raw/master/examples/galaxies/Shen+03_late.txt>`_
# SDSS galaxies which have the following columns: magnitudes, median radius, minus error, and
```

File: examples/galaxies/plot_size.py [hunk 19]
```diff
@@ -0,0 +1,126 @@
...
+#
+# .. math::
+#
+#    \sigma_{ln R} = \sigma_2 + \frac{\sigma_1 - \sigma_2}{1 + 10^{-0.8(M - M_0)}}
+#
+# where :math:`\alpha`, :math:`\beta`, :math:`\gamma`, :math:`\sigma_1`, :math:`\sigma_2` and
+# :math:`M_0` are fitting parameters.
+#
+# In SkyPy, we can sample physical sizes for each galaxy type from lognormal distributions,
+# with median :math:`\bar{R}`, using the functions
```
Comment (rrjbca) [inline examples/galaxies/plot_size.py:42]:
```suggestion
# with median :math:`\bar{R}` and width :math:`\sigma_{ln R}`, using the functions
```

**Review by rrjbca — CHANGES_REQUESTED** [review]

**Review by rrjbca — APPROVED** [review]

**Pull request merged**